src/librustc_middle/ty/context.rs

```diff
@@ -1181,6 +1179,32 @@ impl<'tcx> TyCtxt<'tcx> {
         }
     }
 
+    /// Constructs a `TyKind::Error` type and registers a `delay_span_bug` to ensure it gets used.
+    #[track_caller]
+    pub fn err(self) -> Ty<'tcx> {
+        self.sess.delay_span_bug(
+            DUMMY_SP,
+            &format!(
+                "TyKind::Error constructed but no error reported. {}",
+                std::panic::Location::caller()
+            ),
+        );
+        self.mk_ty(Error)
```

Comment:
Maybe you should add a publicly-unconstructible (outside of this module) `pub struct DelaySpanBugEmitted(());` field to `Error`, even before we change anything about `ErrorReported` itself.

If nothing else, it would catch code that does this (or the const thing, which is probably even more plausible) instead of calling the method.